Support kernel  6.2.9 as in the upcoming Fedora 38

Now Fedora 37 already has kernel `6.2.9-200.fc37.x86_64` as it will be in the upcoming release of Fedora 38.

The build is failing like this:
```
make -C src
make[1]: Entering directory '/home/elastio/elastio-snap/src'
if [ ! -f kernel-config.h ] || tail -1 kernel-config.h | grep -qv '#endif'; then mkdir configure-tests/feature-tests/build; ./genconfig.sh "6.2.9-200.fc37.x86_64" "-w"; fi;
generating configurations for kernel-6.2.9-200.fc37.x86_64
make[2]: Entering directory '/home/elastio/elastio-snap/src/configure-tests/feature-tests'
make[3]: Entering directory '/usr/src/kernels/6.2.9-200.fc37.x86_64'
make[3]: Leaving directory '/usr/src/kernels/6.2.9-200.fc37.x86_64'
make[2]: Leaving directory '/home/elastio/elastio-snap/src/configure-tests/feature-tests'
xargs: warning: options --max-args and --replace/-I/-i are mutually exclusive, ignoring previous --max-args value
performing configure test: HAVE_BDEVNAME - not present
performing configure test: HAVE_ALLOC_DISK - not present
performing configure test: HAVE_BDEV_STACK_LIMITS - not present
performing configure test: HAVE_BDEV_IS_PARTITION - present
performing configure test: HAVE_BDEV_NR_SECTORS - present
performing configure test: HAVE_ADD_DISK_INT - present
performing configure test: HAVE_BDOPS_OPEN_INODE - not present
performing configure test: HAVE_BDOPS_OPEN_INT - not present
performing configure test: HAVE_BDOPS_SUBMIT_BIO_UINT - not present
performing configure test: HAVE_BDOPS_SUBMIT_BIO - present
performing configure test: HAVE_BD_SUPER - present
performing configure test: HAVE_BIO_ALLOC_BIOSET_5 - present
performing configure test: HAVE_BIO_BI_BDEV_BD_DISK - present
performing configure test: HAVE_BIO_BI_REMAINING - not present
performing configure test: HAVE_BIO_BI_BDEV - present
performing configure test: HAVE_BIO_COPY_DEV - not present
performing configure test: HAVE_BIO_BI_POOL - present
performing configure test: HAVE_BIO_ENDIO_INT - not present
performing configure test: HAVE_BIO_ENDIO_1 - present
performing configure test: HAVE_BIOSET_CREATE_3 - not present
performing configure test: HAVE_BIO_FREE_PAGES - present
performing configure test: HAVE_BIO_LIST - present
performing configure test: HAVE_BIO_REMAPPED - present
performing configure test: HAVE_BIO_SET_DEV - present
performing configure test: HAVE_BIOSET_INIT - present
performing configure test: HAVE_BIO_THROTTLED - not present
performing configure test: HAVE_BIO_SPLIT_4 - present
performing configure test: HAVE_BIOSET_NEED_BVECS_FLAG - present
performing configure test: HAVE_BLK_ALLOC_QUEUE_GFP_T - not present
performing configure test: HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID - not present
performing configure test: HAVE_BLKDEV_PUT_1 - not present
performing configure test: HAVE_BLK_MQ_MAKE_REQUEST - not present
performing configure test: HAVE_BLK_MQ_SUBMIT_BIO - not present
performing configure test: HAVE_BLKDEV_GET_BY_PATH - present
performing configure test: HAVE_BLK_SET_STACKING_LIMITS - present
performing configure test: HAVE_BLK_STATUS_T - present
performing configure test: HAVE_BVEC_MERGE_DATA - not present
performing configure test: HAVE_BVEC_ITER_ALL - present
performing configure test: HAVE_BVEC_ITER - present
performing configure test: HAVE_COMPOUND_HEAD - present
performing configure test: HAVE___DENTRY_PATH - not present
performing configure test: HAVE_DENTRY_PATH_RAW - present
performing configure test: HAVE_DISK_LIVE - present
performing configure test: HAVE_ENUM_REQ_OPF - not present
performing configure test: HAVE_D_UNLINKED - present
performing configure test: HAVE_ENUM_REQ_OP - present
performing configure test: HAVE_FILE_INODE - present
performing configure test: HAVE_GENERIC_MAKE_REQUEST - not present
performing configure test: HAVE_FMODE_T - present
performing configure test: HAVE_FOPS_FALLOCATE - present
performing configure test: HAVE_FREEZE_SUPER - present
performing configure test: HAVE_GENHD_FL_NO_PART_SCAN - not present
performing configure test: HAVE_GENHD_FL_NO_PART - present
performing configure test: HAVE_IOPS_FALLOCATE - not present
performing configure test: HAVE_INODE_LOCK - present
performing configure test: HAVE_KERNEL_READ_PPOS - present
performing configure test: HAVE_LOCKS_VERIFY_TRUNCATE - not present
performing configure test: HAVE_KERNEL_WRITE_PPOS - present
performing configure test: HAVE_KERN_PATH - present
performing configure test: HAVE_MAKE_REQUEST_FN_IN_QUEUE - not present
performing configure test: HAVE_MAKE_REQUEST_FN_INT - not present
performing configure test: HAVE_MAKE_REQUEST_FN_VOID - not present
performing configure test: HAVE_MERGE_BVEC_FN - not present
performing configure test: HAVE_MNT_WANT_WRITE - present
performing configure test: HAVE_NOTIFY_CHANGE_2 - not present
performing configure test: HAVE_NOOP_LLSEEK - present
performing configure test: HAVE_NOTIFY_CHANGE_3 - not present
performing configure test: HAVE_PART_NR_SECTS_READ - not present
performing configure test: HAVE_PATH_PUT - present
performing configure test: HAVE_PROC_CREATE_FN_FILE_OPERATIONS - not present
performing configure test: HAVE_PROC_CREATE_FN_PROC_OPS - present
performing configure test: HAVE_SB_START_WRITE - present
performing configure test: HAVE_SI_MEM_AVAILABLE - present
performing configure test: HAVE_STRUCT_PATH - present
performing configure test: HAVE_SUBMIT_BIO_WAIT - not present
performing configure test: HAVE_SUBMIT_BIO_1 - present
performing configure test: HAVE_SYS_OLDUMOUNT - not present
performing configure test: HAVE_TASK_STRUCT_TASK_WORKS_HLIST - not present
performing configure test: HAVE_THAW_BDEV_INT - not present
performing configure test: HAVE_TASK_STRUCT_STATE - present
performing configure test: HAVE_TASK_STRUCT_TASK_WORKS_CB_HEAD - present
performing configure test: HAVE_UAPI_MOUNT_H - present
performing configure test: HAVE_USER_PATH_AT - present
performing configure test: HAVE_VFS_UNLINK_2 - not present
performing configure test: HAVE_VFS_UNLINK_3 - not present
performing configure test: HAVE_UUID_H - present
performing configure test: HAVE_VFS_FALLOCATE - present
performing configure test: HAVE_VZALLOC - present
performing configure test: HAVE_WRITE_ZEROES - present
make[2]: Entering directory '/home/elastio/elastio-snap/src/configure-tests/feature-tests'
make[3]: Entering directory '/usr/src/kernels/6.2.9-200.fc37.x86_64'
make[3]: Leaving directory '/usr/src/kernels/6.2.9-200.fc37.x86_64'
make[2]: Leaving directory '/home/elastio/elastio-snap/src/configure-tests/feature-tests'
performing sys_mount lookup
performing sys_umount lookup
performing __x64_sys_mount lookup
performing __x64_sys_umount lookup
performing __arm64_sys_mount lookup
performing __arm64_sys_umount lookup
performing __change_memory_common lookup
performing sys_oldumount lookup
performing sys_call_table lookup
performing kfree lookup
performing blk_mq_submit_bio lookup
performing blk_alloc_queue lookup
performing get_super lookup
checking CONFIG_ARM64_PAGE_SHIFT
make -C /lib/modules/6.2.9-200.fc37.x86_64/build M=/home/elastio/elastio-snap/src modules
make[2]: Entering directory '/usr/src/kernels/6.2.9-200.fc37.x86_64'
  CC [M]  /home/elastio/elastio-snap/src/elastio-snap.o
/home/elastio/elastio-snap/src/elastio-snap.c: In function ‘elastio_snap_set_bio_ops’:
/home/elastio/elastio-snap/src/elastio-snap.c:300:9: error: implicit declaration of function ‘bio_set_op_attrs’ [-Werror=implicit-function-declaration]
  300 |         bio_set_op_attrs(bio, op, op_flags);
      |         ^~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:252: /home/elastio/elastio-snap/src/elastio-snap.o] Error 1
make[2]: *** [Makefile:2025: /home/elastio/elastio-snap/src] Error 2
make[2]: Leaving directory '/usr/src/kernels/6.2.9-200.fc37.x86_64'
make[1]: *** [Makefile:14: default] Error 2
make[1]: Leaving directory '/home/elastio/elastio-snap/src'
make: *** [Makefile:32: driver] Error 2
Error: Process completed with exit code 2.
```

--- a/.github/buildbox/Vagrantfile
+++ b/.github/buildbox/Vagrantfile
@@ -38,14 +38,14 @@ DISTRO['amd64'] = [ "debian8", "debian9", "debian10", "debian11",
                     "amazon2", "amazon2023",
                     "centos7", "centos8", "centos9",
                     "alma8", "alma9",
-                    "fedora31", "fedora32", "fedora34", "fedora35", "fedora36", "fedora37",
+                    "fedora31", "fedora32", "fedora34", "fedora35", "fedora36", "fedora37", "fedora38",
                     "ubuntu2004", "ubuntu2204" ]
 
 DISTRO['arm64'] = [ "debian10", "debian11",
                     "amazon2", "amazon2023",
                     "centos8", "centos9",
                     "alma8", "alma9",
-                    "fedora35", "fedora36", "fedora37",
+                    "fedora35", "fedora36", "fedora37", "fedora38",
                     "ubuntu2204" ]
 
 # NOTE: This environment variable should be set to enable triggers of the type "action".

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -33,7 +33,7 @@ jobs:
           amazon2, amazon2023,
           centos7, centos8, centos9,
           debian8, debian9, debian10, debian11,
-          fedora31, fedora32, fedora34, fedora35, fedora36, fedora37,
+          fedora31, fedora32, fedora34, fedora35, fedora36, fedora37, fedora38,
           ubuntu2004, ubuntu2204
         ]
         arch: [ amd64 ]
@@ -53,6 +53,8 @@ jobs:
           - distro: fedora36
             arch: arm64
           - distro: fedora37
+            arch: arm64
+          - distro: fedora38
             arch: arm64
           - distro: ubuntu2204
             arch: arm64
@@ -186,15 +188,15 @@ jobs:
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -f $fs"
           done
         working-directory: ${{env.BOX_DIR}}
-        # For now tests are taking 10-20 seconds per 1 fs on amd64 and about 2 minutes on arm64. But they can hang.
-        # 10 minutes seems to be reasonable timeout.
-        timeout-minutes: 10
+        # For now tests are taking up to 40 seconds per 1 fs on amd64 and about 2 minutes on arm64 or older 3.x kernels. But they can hang.
+        # 20 minutes seems to be reasonable timeout.
+        timeout-minutes: 20
 
       - name: Show debug information
         if: always()
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cat tests/dmesg.log; sudo dmesg -c; lsmod"
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 1
+        timeout-minutes: 2
 
       - name: Run tests on LVM (loop device)
         if: "${{ matrix.distro != 'debian8'}}"
@@ -204,14 +206,14 @@ jobs:
           done
         working-directory: ${{env.BOX_DIR}}
         # For now tests are taking 10-20 seconds per 1 fs on amd64 and about 2 minutes on arm64. But they can hang.
-        # 10 minutes seems to be reasonable timeout.
-        timeout-minutes: 10
+        # 20 minutes seems to be reasonable timeout.
+        timeout-minutes: 20
 
       - name: Show debug information
         if: always()
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cat tests/dmesg.log; sudo dmesg -c; lsmod"
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 1
+        timeout-minutes: 2
 
       - name: Run tests on RAID (loop device)
         if: "${{ matrix.distro != 'debian8'}}"
@@ -220,9 +222,9 @@ jobs:
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -f $fs --raid"
           done
         working-directory: ${{env.BOX_DIR}}
-        # For now tests are taking 10-20 seconds per 1 fs on amd64 and about 2 minutes on arm64. But they can hang.
-        # 10 minutes seems to be reasonable timeout.
-        timeout-minutes: 10
+        # For now tests are taking up to 40 seconds per 1 fs on amd64 and about 2 minutes on arm64 or older 3.x kernels. But they can hang.
+        # 20 minutes seems to be reasonable timeout.
+        timeout-minutes: 20
 
       - name: Show debug information
         if: always()
@@ -232,10 +234,10 @@ jobs:
 
       - name: Attach qcow2 disks
         run: |
+          SZ="2200M"
           ARCH=$(uname -m)
           TEST_IMAGES=(${TEST_IMAGES})
           TEST_DRIVES=(${TEST_DRIVES})
-          [ ${ARCH} == "x86_64" ] && SZ="1G" || SZ="256M"
           [ ${ARCH} != "x86_64" ] && VIRSH_FLAGS="--config" || true
           for i in ${!TEST_IMAGES[*]}; do
             qemu-img create -f qcow2 ${TEST_IMAGES[i]} $SZ
@@ -261,13 +263,13 @@ jobs:
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -d /dev/vdb1 -f $fs"
           done
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 10
+        timeout-minutes: 25
 
       - name: Show debug information
         if: always()
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cat tests/dmesg.log; sudo dmesg -c; lsmod"
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 1
+        timeout-minutes: 2
 
       - name: Run tests on LVM (qcow2 disks)
         if: "${{ matrix.distro != 'debian8'}}"
@@ -276,13 +278,13 @@ jobs:
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -d /dev/vdb -d /dev/vdc -f $fs --lvm"
           done
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 15
+        timeout-minutes: 25
 
       - name: Show debug information
         if: always()
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cat tests/dmesg.log; sudo dmesg -c; lsmod"
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 1
+        timeout-minutes: 2
 
       - name: Run tests on RAID (qcow2 disks)
         if: "${{ matrix.distro != 'debian8'}}"
@@ -294,13 +296,13 @@ jobs:
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -d /dev/vdb -d /dev/vdc -f $fs --raid"
           done
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 15
+        timeout-minutes: 25
 
       - name: Show debug information
         if: always()
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cat tests/dmesg.log; sudo dmesg -c; lsmod"
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 1
+        timeout-minutes: 2
 
       - name: Run tests multipart (qcow2 disk)
         run: |
@@ -308,13 +310,13 @@ jobs:
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -d /dev/vdb -f $fs -t test_multipart"
           done
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 10
+        timeout-minutes: 15
 
       - name: Show debug information
         if: always()
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cat tests/dmesg.log; sudo dmesg -c; lsmod"
         working-directory: ${{env.BOX_DIR}}
-        timeout-minutes: 1
+        timeout-minutes: 2
 
       - name: Detach external drives
         if: always()

--- a/src/configure-tests/feature-tests/bio_set_op_attrs.c
+++ b/src/configure-tests/feature-tests/bio_set_op_attrs.c
@@ -1,0 +1,16 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2023 Elastio Software Inc.
+ */
+
+// kernel_version < 6.2
+
+#include "includes.h"
+MODULE_LICENSE("GPL");
+
+static inline void dummy(void){
+	struct bio bio;
+
+	bio_set_op_attrs(&bio, REQ_OP_READ, 0);
+}

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -298,8 +298,12 @@ typedef enum req_opf req_op_t;
 #endif
 
 static inline void elastio_snap_set_bio_ops(struct bio *bio, req_op_t op, unsigned op_flags){
+#ifdef HAVE_BIO_SET_OP_ATTRS
 	bio->bi_opf = 0;
 	bio_set_op_attrs(bio, op, op_flags);
+#else
+	bio->bi_opf = op | op_flags;
+#endif
 }
 
 static inline int elastio_snap_bio_op_flagged(struct bio *bio, unsigned int flag){
@@ -3341,6 +3345,7 @@ static inline struct inode *page_get_inode(struct page *pg){
 #endif
 	if(PageAnon(pg)) return NULL;
 	if(!pg->mapping) return NULL;
+	if (!virt_addr_valid(pg->mapping)) return NULL;
 	return pg->mapping->host;
 }
 

--- a/tests/devicetestcase.py
+++ b/tests/devicetestcase.py
@@ -22,7 +22,7 @@ class DeviceTestCase(unittest.TestCase):
         seeds = []
         cls.backing_stores = []
         cls.devices = []
-        cls.size_mb = 256
+        cls.size_mb = 512
         cls.is_raid = False
 
         cls.kmod = kmod.Module("../src/elastio-snap.ko")

--- a/tests/devicetestcase_multipart.py
+++ b/tests/devicetestcase_multipart.py
@@ -25,7 +25,7 @@ class DeviceTestCaseMultipart(unittest.TestCase):
 
         # We create the device big enough to perform
         # representative 'test_multipart_modify_origins'
-        cls.size_mb = 1024
+        cls.size_mb = 2200
 
         for i in range(cls.part_count):
             # Unexpectedly randint can generate 2 same numbers in a row.

--- a/tests/test_multipart.py
+++ b/tests/test_multipart.py
@@ -93,7 +93,7 @@ class TestMultipart(DeviceTestCaseMultipart):
 
             # We subtract a couple of megabytes to make sure the cow
             # file won't overflow during the test
-            file_size_mb = math.floor(dev_size_mb * 0.1) - 2
+            file_size_mb = math.floor(dev_size_mb * 0.1) - 20
 
             testfile = "{}/testfile".format(self.mounts[i])
             snapfile = "{}/testfile".format(self.snap_mounts[i])

--- a/tests/test_snapshot.py
+++ b/tests/test_snapshot.py
@@ -42,7 +42,7 @@ class TestSnapshot(DeviceTestCase):
         if self.is_raid == True:
             file_size_mb = math.floor(dev_size_mb * 0.06)
         else:
-            file_size_mb = math.floor(dev_size_mb * 0.1) - 5
+            file_size_mb = math.floor(dev_size_mb * 0.1) - 20
 
         testfile = "{}/testfile".format(self.mount)
         snapfile = "{}/testfile".format(self.snap_mount)

--- a/tests/util.py
+++ b/tests/util.py
@@ -24,7 +24,7 @@ def unmount(path, retry_on_dev_busy=True):
     cmd = ["umount", path]
     # subprocess.run is introduced in Python 3.5
     if not retry_on_dev_busy or sys.version_info <= (3, 5):
-        subprocess.check_call(cmd, timeout=10)
+        subprocess.check_call(cmd, timeout=20)
     else:
         # The retries on device busy error are necessary on Ubuntu 22.04, kernel 5.15
         # for the tests test_destroy_unverified_incremental and test_destroy_unverified_snapshot.
@@ -48,7 +48,7 @@ def dd(ifile, ofile, count, **kwargs):
     for k, v in kwargs.items():
         cmd.append("{}={}".format(k, v))
 
-    subprocess.check_call(cmd, timeout=60)
+    subprocess.check_call(cmd, timeout=240)
 
 
 def md5sum(path):

